src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs

```diff
@@ -149,83 +150,110 @@ public static partial class JsonSerializer
         }
 
         // If this method is changed, also change ApplyValueToEnumerable.
-        internal static void ApplyObjectToEnumerable(object value, JsonSerializerOptions options, ref ReadStackFrame frame, bool setPropertyDirectly = false)
+        internal static void ApplyObjectToEnumerable(
+            object value,
+            JsonSerializerOptions options,
+            ref ReadStack state,
+            ref Utf8JsonReader reader,
+            bool setPropertyDirectly = false)
         {
-            if (frame.IsEnumerable)
+            if (state.Current.IsEnumerable)
             {
-                if (frame.TempEnumerableValues != null)
+                if (state.Current.TempEnumerableValues != null)
                 {
-                    frame.TempEnumerableValues.Add(value);
+                    state.Current.TempEnumerableValues.Add(value);
                 }
                 else
                 {
-                    ((IList)frame.ReturnValue).Add(value);
+                    ((IList)state.Current.ReturnValue).Add(value);
                 }
             }
-            else if (!setPropertyDirectly && frame.IsPropertyEnumerable)
+            else if (!setPropertyDirectly && state.Current.IsPropertyEnumerable)
             {
-                Debug.Assert(frame.JsonPropertyInfo != null);
-                Debug.Assert(frame.ReturnValue != null);
-                if (frame.TempEnumerableValues != null)
+                Debug.Assert(state.Current.JsonPropertyInfo != null);
+                Debug.Assert(state.Current.ReturnValue != null);
+                if (state.Current.TempEnumerableValues != null)
                 {
-                    frame.TempEnumerableValues.Add(value);
+                    state.Current.TempEnumerableValues.Add(value);
                 }
                 else
                 {
-                    ((IList)frame.JsonPropertyInfo.GetValueAsObject(frame.ReturnValue)).Add(value);
+                    ((IList)state.Current.JsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue)).Add(value);
                 }
             }
-            else if (frame.IsDictionary)
+            else if (state.Current.IsDictionary)
             {
-                Debug.Assert(frame.ReturnValue != null);
-                ((IDictionary)frame.JsonPropertyInfo.GetValueAsObject(frame.ReturnValue)).Add(frame.KeyName, value);
+                Debug.Assert(state.Current.ReturnValue != null);
+                IDictionary dictionary = (IDictionary)state.Current.JsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue);
```

Comment:
Will this cast always succeed?

Reply:
Yes. We verified earlier that to be a "dictionary" it has to either implement `IDictionary<,>` or `IReadOnlyDictionary<,>` in which case we will instantiate a `Dictionary<,>`.

Reply:
Can you please add that as a comment.